blockchain/blockchain.go

```diff
@@ -1584,6 +1584,43 @@ func (bc *BlockChain) insertChain(chain types.Blocks) (int, []interface{}, []*ty
 
 	// Iterate over the blocks and insert when the verifier permits
 	for i, block := range chain {
+		// Create a new trie using the parent block and report an
+		// error if it fails.
+		var parent *types.Block
+		if i == 0 {
+			parent = bc.GetBlock(block.ParentHash(), block.NumberU64()-1)
+		} else {
+			parent = chain[i-1]
+		}
+
+		// If we have a followup block, run that against the current state to pre-cache
+		// transactions and probabilistically some of the account/storage trie nodes.
+		var followupInterrupt uint32
+
+		if bc.cacheConfig.TrieNodeCacheConfig.NumFetcherPrefetchWorker > 0 {
+			// Tx prefetcher is enabled for all cases (both single and multiple block insertion).
+			for ti := range block.Transactions() {
+				select {
+				case bc.prefetchTxCh <- prefetchTx{ti, block, &followupInterrupt}:
+				default:
+				}
+			}
+			if i < len(chain)-1 {
+				// current block is not the last one, so prefetch the right next block
+				followup := chain[i+1]
+				go func(start time.Time) {
+					throwaway, _ := state.NewForPrefetching(parent.Root(), bc.stateCache)
+					vmCfg := bc.vmConfig
+					vmCfg.Prefetching = true
+					bc.prefetcher.Prefetch(followup, throwaway, vmCfg, &followupInterrupt)
+
+					blockPrefetchExecuteTimer.Update(time.Since(start))
+					if atomic.LoadUint32(&followupInterrupt) == 1 {
+						blockPrefetchInterruptMeter.Mark(1)
+					}
+				}(time.Now())
+			}
+		}
 		// If the chain is terminating, stop processing blocks
 		if atomic.LoadInt32(&bc.procInterrupt) == 1 {
 			logger.Debug("Premature abort during blocks processing")
```

Comment:
@kjhman21 Is it better to place the prefetching logic below of this `if` branch? It's minor, but `blockchain` needs to wait to finish prefetching when it is closing.

Reply:
@aidan-kwon Hmm.. Actually prefetch logic just passes data to another goroutine, so I think it does not take up much time. Do you think it would be better even if we are using goroutine?

Reply:
@kjhman21 Oh, yeah. I missed prefetching runs concurrently. Let' me check another thing, concurrency issues...
- Prefetching logic can read Database and TrieNodeCache after Blockchain is stopped. It would be fine. Database and TrieNodeCache don't panic even though read operation is called after the close. 
- `bc.prefechTxCh` can receives items after the channel is closed. It causes panic. It is an existing issue, not caused by this change. However, this change (checking  `bc.procInterrupt` later) may increase the possibility of panic. @kjhman21 Could you review this comment?

Reply:
@aidan-kwon 
Okay. So, you think it would be better to check `bc.procInterrupt` first. Right?

Reply:
@kjhman21 Yes

Reply:
@aidan-kwon All right, then. :)

Reply:
@kjhman21 Thanks you!